File: ios/TimberCruisingApp/Persistence/Repositories/PlannedPlotRepository.swift
// Spec §8 + REQ-PRJ-004, REQ-NAV-001 (visited vs remaining).

import Foundation
import CoreData
import Models

public protocol PlannedPlotRepository {
    func create(_ p: PlannedPlot) throws -> PlannedPlot
    func read(id: UUID) throws -> PlannedPlot?
    func update(_ p: PlannedPlot) throws -> PlannedPlot
    func delete(id: UUID) throws
    func listByProject(_ projectId: UUID) throws -> [PlannedPlot]
    func listUnvisited(projectId: UUID) throws -> [PlannedPlot]
}

public final class CoreDataPlannedPlotRepository: PlannedPlotRepository {
    private let stack: CoreDataStack
    public init(stack: CoreDataStack) { self.stack = stack }

    public func create(_ p: PlannedPlot) throws -> PlannedPlot {
        try performWrite(stack: stack) { ctx in
            let e = try insert(PlannedPlotEntity.self, entityName: "PlannedPlotEntity", in: ctx)
            PlannedPlotMapper.apply(p, to: e)
            return p
        }
    }

    public func read(id: UUID) throws -> PlannedPlot? {
        try performRead(stack: stack) { ctx in
            guard let e = try fetchOne(PlannedPlotEntity.self, entityName: "PlannedPlotEntity",
                                       keyPath: "id", value: id as CVarArg, in: ctx)
            else { return nil }
            return PlannedPlotMapper.toStruct(e)
        }
    }

    public func update(_ p: PlannedPlot) throws -> PlannedPlot {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(PlannedPlotEntity.self, entityName: "PlannedPlotEntity",
                                       keyPath: "id", value: p.id as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: p.id.uuidString) }
            PlannedPlotMapper.apply(p, to: e)
            return p
        }
    }

    public func delete(id: UUID) throws {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(PlannedPlotEntity.self, entityName: "PlannedPlotEntity",
                                       keyPath: "id", value: id as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: id.uuidString) }
            ctx.delete(e)
        }
    }

    public func listByProject(_ projectId: UUID) throws -> [PlannedPlot] {
        try performRead(stack: stack) { ctx in
            let pred = NSPredicate(format: "projectId == %@", projectId as CVarArg)
            let sort = [NSSortDescriptor(key: "plotNumber", ascending: true)]
            return try fetchMany(PlannedPlotEntity.self, entityName: "PlannedPlotEntity",
                                 predicate: pred, sort: sort, in: ctx)
                .map(PlannedPlotMapper.toStruct)
        }
    }

    public func listUnvisited(projectId: UUID) throws -> [PlannedPlot] {
        try performRead(stack: stack) { ctx in
            let pred = NSPredicate(format: "projectId == %@ AND visited == NO", projectId as CVarArg)
            let sort = [NSSortDescriptor(key: "plotNumber", ascending: true)]
            return try fetchMany(PlannedPlotEntity.self, entityName: "PlannedPlotEntity",
                                 predicate: pred, sort: sort, in: ctx)
                .map(PlannedPlotMapper.toStruct)
        }
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/data/cruise/PlannedPlotRepository.kt
// Port of iOS Persistence/Repositories/PlannedPlotRepository.swift.
// Spec §8 + REQ-PRJ-004, REQ-NAV-001 (visited vs remaining).

package com.hcjeong.forestix.data.cruise

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.util.UUID

interface PlannedPlotRepository {
    suspend fun create(p: PlannedPlot): PlannedPlot
    suspend fun read(id: UUID): PlannedPlot?
    suspend fun update(p: PlannedPlot): PlannedPlot
    suspend fun delete(id: UUID)
    suspend fun listByProject(projectId: UUID): List<PlannedPlot>
    suspend fun listUnvisited(projectId: UUID): List<PlannedPlot>

    /// Live view of `listByProject()` for Compose screens.
    fun observeByProject(projectId: UUID): Flow<List<PlannedPlot>>
}

class RoomPlannedPlotRepository(private val dao: PlannedPlotDao) : PlannedPlotRepository {

    override suspend fun create(p: PlannedPlot): PlannedPlot {
        dao.upsert(PlannedPlotMapper.apply(p))
        return p
    }

    override suspend fun read(id: UUID): PlannedPlot? =
        dao.byId(id)?.let(PlannedPlotMapper::toStruct)

    override suspend fun update(p: PlannedPlot): PlannedPlot {
        dao.byId(p.id) ?: throw CruiseDataError.NotFound(p.id.toString())
        dao.upsert(PlannedPlotMapper.apply(p))
        return p
    }

    override suspend fun delete(id: UUID) {
        if (dao.deleteById(id) == 0) throw CruiseDataError.NotFound(id.toString())
    }

    override suspend fun listByProject(projectId: UUID): List<PlannedPlot> =
        dao.listByProject(projectId).map(PlannedPlotMapper::toStruct)

    override suspend fun listUnvisited(projectId: UUID): List<PlannedPlot> =
        dao.listUnvisited(projectId).map(PlannedPlotMapper::toStruct)

    override fun observeByProject(projectId: UUID): Flow<List<PlannedPlot>> =
        dao.observeByProject(projectId).map { rows -> rows.map(PlannedPlotMapper::toStruct) }
}
